odoo/addons/base/models/res_partner.py

```diff
@@ -517,7 +517,14 @@ def write(self, vals):
             self.invalidate_cache(['user_ids'], self._ids)
             for partner in self:
                 if partner.active and partner.user_ids:
-                    raise ValidationError(_('You cannot archive a contact linked to an internal user.'))
+                    if self.env['res.users'].sudo(False).browse(partner.user_ids.id).check_access_rights('write', raise_exception=False):
```

Comment:
```suggestion
                    if self.env['res.users'].sudo(False).check_access_rights('write', raise_exception=False):
```
`check_access_rights` checks on the model, not the record. If you want to check on this specific record (e.g. multicompany), the method is `.check_access_rule`. The first one is probably enough for your scenario.

Reply:
Complementary note: contrary to what you might expect `check_access_rule` does not imply or call `check_access_rights`, so you should always have the latter, and optionally the former if you want to check for access on specific records and not only the model in general.

Also `check_access_rule` doesn't have a `raise_exception`, so the usual pattern is:
```python
try:
    records.check_access_rights(operation)
    records.check_access_rule(operation)
except AccessError:
    "user doesn't have access"
else:
    "user has access"
```
(depending on the specifics, the contents of the `else` block may be folded into the `try` block, or live after the entire `try/except` statement).